In [142]:
import numpy as np
import pandas as pd

In [143]:
hawks_df = pd.read_csv('./hawks.csv')
hawks_df.head(10)

Month  Day  Year CaptureTime ReleaseTime   BandNumber Species Age  Sex  \
0      9   19  1992       13:30         NaN    877-76317      RT   I  NaN   
1      9   22  1992       10:30                877-76318      RT   I  NaN   
2      9   23  1992       12:45                877-76319      RT   I  NaN   
3      9   23  1992       10:50                745-49508      CH   I    F   
4      9   27  1992       11:15               1253-98801      SS   I    F   
5      9   28  1992       11:25              1207-55910       RT   I  NaN   
6      9   28  1992       13:30                877-76320      RT   I  NaN   
7      9   29  1992       11:45                877-76321      RT   A  NaN   
8      9   29  1992       15:35                877-76322      RT   A  NaN   
9      9   30  1992       13:45               1207-55911      RT   I  NaN   

    Wing  Weight  Culmen  Hallux  Tail  StandardTail  Tarsus  WingPitFat  \
0  385.0   920.0    25.7    30.1   219           NaN     NaN         NaN   
1  376.0   930.0     NaN     NaN   221           NaN     NaN         NaN   
2  381.0   990.0    26.7    31.3   235           NaN     NaN         NaN   
3  265.0   470.0    18.7    23.5   220           NaN     NaN         NaN   
4  205.0   170.0    12.5    14.3   157           NaN     NaN         NaN   
5  412.0  1090.0    28.5    32.2   230           NaN     NaN         NaN   
6  370.0   960.0    25.3    30.1   212           NaN     NaN         NaN   
7  375.0   855.0    27.2    30.0   243           NaN     NaN         NaN   
8  412.0  1210.0    29.3    31.3   210           NaN     NaN         NaN   
9  405.0  1120.0    26.0    30.2   238           NaN     NaN         NaN   

   KeelFat  Crop  
0      NaN   NaN  
1      NaN   NaN  
2      NaN   NaN  
3      NaN   NaN  
4      NaN   NaN  
5      NaN   NaN  
6      NaN   NaN  
7      NaN   NaN  
8      NaN   NaN  
9      NaN   NaN

In [144]:
hawks_df.shape

(908, 19)

In [145]:
def check(df):
    for item in df.columns:
        print(df[item].dtype)
        print('<' + item + '>', ((df[item].isna()) | (df[item] == ' ')).value_counts(normalize=True)*100)
        print()

# kiểm tra về kiểu dữ liệu và các dữ liệu bị thiếu trong df 
check(hawks_df)

int64
<Month> False    100.0
Name: Month, dtype: float64

int64
<Day> False    100.0
Name: Day, dtype: float64

int64
<Year> False    100.0
Name: Year, dtype: float64

object
<CaptureTime> False    99.889868
True      0.110132
Name: CaptureTime, dtype: float64

object
<ReleaseTime> True     92.84141
False     7.15859
Name: ReleaseTime, dtype: float64

object
<BandNumber> False    99.779736
True      0.220264
Name: BandNumber, dtype: float64

object
<Species> False    100.0
Name: Species, dtype: float64

object
<Age> False    100.0
Name: Age, dtype: float64

object
<Sex> True     63.436123
False    36.563877
Name: Sex, dtype: float64

float64
<Wing> False    99.889868
True      0.110132
Name: Wing, dtype: float64

float64
<Weight> False    98.898678
True      1.101322
Name: Weight, dtype: float64

float64
<Culmen> False    99.229075
True      0.770925
Name: Culmen, dtype: float64

float64
<Hallux> False    99.339207
True      0.660793
Name: Hallux, dtype: float64

int64
<Tail> False    

- Điền đầy dữ liệu số bằng mean và điền đầy dữ liệu định danh bằng mode  
- Đối với những thuộc tính bị thiếu trên 50% thì xoá luôn thuộc tính đi  
- Do đó các thuộc tính bị xoá sẽ là
    - ReleaseTime: thiếu 92.8%
    - Sex: thiếu 63.4%
    - Tarsus: thiếu 91.7%
    - WingPitFat: thiếu 91.5%
- Các thuộc tính sẽ được điền đầy bằng phương pháp mean
    - Wing 
    - Weight
    - Culmen 
    - Hallux 
    - StandardTail
    - KeelFat 
    - Crop
- Thuộc tính sẽ được điền đầy bằng phương pháp mode: CaptureTime 
- Quan sát dữ liệu cột BandNumber, ta thấy thuộc tính này hoàn toàn có thể là thuộc tính index vì tất cả các dòng dữ liệu đều khác nhau. Do đó, ta xoá luôn cột này

In [146]:
# xoá cột
hawks_df.drop(columns=['ReleaseTime', 'Sex', 'Tarsus', 'WingPitFat', 'BandNumber'], inplace=True)
hawks_df.head()

Month  Day  Year CaptureTime Species Age   Wing  Weight  Culmen  Hallux  \
0      9   19  1992       13:30      RT   I  385.0   920.0    25.7    30.1   
1      9   22  1992       10:30      RT   I  376.0   930.0     NaN     NaN   
2      9   23  1992       12:45      RT   I  381.0   990.0    26.7    31.3   
3      9   23  1992       10:50      CH   I  265.0   470.0    18.7    23.5   
4      9   27  1992       11:15      SS   I  205.0   170.0    12.5    14.3   

   Tail  StandardTail  KeelFat  Crop  
0   219           NaN      NaN   NaN  
1   221           NaN      NaN   NaN  
2   235           NaN      NaN   NaN  
3   220           NaN      NaN   NaN  
4   157           NaN      NaN   NaN

Tại các cột dữ liệu định danh, có 2 cột 

In [147]:
# điền đầy các cột numeric bằng mean 
num_col = ['Wing' ,'Weight','Culmen' ,'Hallux' ,'StandardTail','KeelFat' ,'Crop']

for item in num_col:
    hawks_df[item] = hawks_df[item].fillna(hawks_df[item].mean())

hawks_df.head()

Month  Day  Year CaptureTime Species Age   Wing  Weight     Culmen  \
0      9   19  1992       13:30      RT   I  385.0   920.0  25.700000   
1      9   22  1992       10:30      RT   I  376.0   930.0  21.801498   
2      9   23  1992       12:45      RT   I  381.0   990.0  26.700000   
3      9   23  1992       10:50      CH   I  265.0   470.0  18.700000   
4      9   27  1992       11:15      SS   I  205.0   170.0  12.500000   

      Hallux  Tail  StandardTail   KeelFat     Crop  
0  30.100000   219    199.182137  2.184303  0.23454  
1  26.410865   221    199.182137  2.184303  0.23454  
2  31.300000   235    199.182137  2.184303  0.23454  
3  23.500000   220    199.182137  2.184303  0.23454  
4  14.300000   157    199.182137  2.184303  0.23454

Đối với các cột norminal bị thiếu dữ liệu, cột `CaptureTime`, `BandNumber`


In [148]:
mode = hawks_df['CaptureTime'].mode()
hawks_df['CaptureTime'] = hawks_df['CaptureTime'].replace([np.nan, ' '], mode)
check(hawks_df)

int64
<Month> False    100.0
Name: Month, dtype: float64

int64
<Day> False    100.0
Name: Day, dtype: float64

int64
<Year> False    100.0
Name: Year, dtype: float64

object
<CaptureTime> False    100.0
Name: CaptureTime, dtype: float64

object
<Species> False    100.0
Name: Species, dtype: float64

object
<Age> False    100.0
Name: Age, dtype: float64

float64
<Wing> False    100.0
Name: Wing, dtype: float64

float64
<Weight> False    100.0
Name: Weight, dtype: float64

float64
<Culmen> False    100.0
Name: Culmen, dtype: float64

float64
<Hallux> False    100.0
Name: Hallux, dtype: float64

int64
<Tail> False    100.0
Name: Tail, dtype: float64

float64
<StandardTail> False    100.0
Name: StandardTail, dtype: float64

float64
<KeelFat> False    100.0
Name: KeelFat, dtype: float64

float64
<Crop> False    100.0
Name: Crop, dtype: float64



In [151]:
# lưu file
hawks_df.to_csv('afterPreprocess.csv', header=True, index=False)